chain/sync.go

```diff
@@ -102,6 +108,18 @@ func (s *Syncer) DisableDiscoverAccounts() {
 	s.mu.Unlock()
 }
 
+// Synced returns whether the syncer has completed syncing to the backend and
+// the target height it is attempting to sync to.
+func (s *Syncer) Synced(ctx context.Context) (bool, int32) {
```

Comment:
This api seems like it could be creating confusing results at times.  The syncer will mark itself synced once at startup, but as new headers are announced, the second return value for the height being synced to will continue to increase.  furthermore, it might show as being synced, but a just-announced header hasn't been fully processed yet, while the wallet is currently synced to a point behind this header.

Is there a reason we can't split up this method into two, with each returning their respective result?

Reply:
That would just move the issue to callers, because both values are used in the same spot (i.e. the `syncstatus` jsonrpc handler). It already had the issues you outline before this PR, except it was using 2 functions (`Synced()` and `EstimateMainChainTip()`).

The `synced` flag is also never cleared now (neither in the SPV or RPC syncers), so the wallet can end up in a situation where it's not connected to any peers for days and yet the synced flag is remains true.

If we decide on what semantics we want from this synced flag, then I can do it in either 1 or 2 functions (though depending on the specific necessity for locking, doing as 2 functions could introduce a hard-to-hit in practice race condition).

We could also modify `SyncStatusResult` to be more granular and return both the target sync height, the current mainchain tip (height, hash and timestamp), and whether the initial rescan and addr discovery have completed, and leave up to the caller to decide whether the wallet should be considered synced.

Reply:
Hmm, I think these are the semantics I'd want:

If the syncer is still in initial sync, synced=false and height=whatever we are currently syncing to.

If the syncer has finished initial sync, as long as the rpc client remains connected (or in spv mode, as long as at least one peer remains connected) we keep returning synced=true.  But instead of returning the height of unprocessed announced blocks, we use the block height we have most recently processed completely.

Reply:
> We could also modify SyncStatusResult to be more granular and return both the target sync height, the current mainchain tip (height, hash and timestamp), and whether the initial rescan and addr discovery have completed, and leave up to the caller to decide whether the wallet should be considered synced.

Yeah, agreeing with this idea.  syncstatus should more or less pass through the 'synced' result given the above semantics.  And then also include the block height and/or hash of both the most recently fully processed block, as well as the most recent one announced by dcrd or a spv peer.

Reply:
Ok, the [next commit](https://github.com/decred/dcrwallet/pull/2320/commits/295685884bef8c2a8f3aca375397da12306624b0) should make the syncers follow those semantics (mainchain tip if synced, otherwise the target height).